viewer/Viewer.cpp

```diff
@@ -750,25 +842,138 @@ namespace Enki
+		QMatrix4x4 transformMatrix = (projection*modelview).inverted();
+
+		// cursor position in viewport coordinates
+		float fragmentX = (float)(cursorPosition.x() - width()/2)/(width()/2);
+		float fragmentY = (float)(height() - cursorPosition.y() - height()/2)/(height()/2);
+		float depth;
+		glReadPixels( cursorPosition.x(), height() - cursorPosition.y(), 1, 1, GL_DEPTH_COMPONENT, GL_FLOAT, &depth );
+
+		QVector4D input(fragmentX, fragmentY, 2*depth - 1, 1);
+		input = transformMatrix*input;
+
+		if (input.w() != 0.0) // valid pointed point
+		{
+			pointedPoint = QVector3D(input.x(),input.y(),input.z());
+			pointedPoint /= input.w();
+		}
+		else return;
+
+		// prepare to find which object is pointed
+		Point cursor2Dpoint(pointedPoint.x(),pointedPoint.y());
+		float cursorRadius = 0.05f;
+		for (World::ObjectsIterator it = world->objects.begin(); it != world->objects.end(); ++it)
+		{
+			const Vector distOCtoOC = (*it)->pos - cursor2Dpoint;		// distance between object bounding circle center and pointed point
+			const double addedRay = (*it)->getRadius() + cursorRadius;	// sum of bounded circle radius
+			if (distOCtoOC.norm2() <= (addedRay*addedRay)) 			// cursor point colide bounding circle
+			{
+				if(!(*it)->getHull().empty())				// check pointer circle and bject hull
+				{
+					PhysicalObject::Hull hull = (*it)->getHull();
+					for (PhysicalObject::Hull::const_iterator it2 = hull.begin(); it2 != hull.end(); ++it2) // check all convex shape of hull
+					{
+						const Polygone shape = it2->getTransformedShape();
+						unsigned int inside = 0;
+
+						// standard test : if circularObject is inside a convex shape
+						for (unsigned int i=0; i<shape.size(); i++)
+						{
+							const size_t next=(i+1)%shape.size();
+							const Segment s(shape[i].x, shape[i].y, shape[next].x, shape[next].y);
+							const double d = s.dist(cursor2Dpoint);
+
+							if(d<0 && std::abs(d)>cursorRadius) // out of hull
+								break;
+							else inside++;
+						}
+						if(inside == shape.size()) // inside of hull
+						{
+							pointedObject = *it;
+							break;
+						}
+					}
+				}
+				else	// object circle collide cursor circle => test already done !
+					pointedObject = *it;
+			}
+		}
+	}
+	void ViewerWidget::displayMessages()
+	{
+		while(messageList.size() > 20)
+			messageList.pop_front();
+
+		unsigned int origin = clamp((int)(height() - (messageList.size()-1)*15)-5, 0, height()-5);
+		unsigned int i = 0;
+		for(std::list<viewerMessage>::iterator it = messageList.begin(); it != messageList.end();i++)
+		{
+			glColor4d(0,0,0,clamp(it->second/30.,0.,1.));
+			renderText(5,origin + i*15,it->first);
+
+			if(it->second)
+			{
+				it->second--;
+				++it;
+			}
+			else it = messageList.erase(it);
+		}
+	}
+
+	void ViewerWidget::paintGL()
+	{
+		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+		float znear = 0.5;
+		if(trackballView && selectedObject)
```

Comment:
Enki's coding convention is to put a space between `if` and the opening parenthesis: `if (`.